build-tools-internal/src/main/java/org/elasticsearch/gradle/internal/transport/ValidateTransportVersionDefinitionsTask.java

```diff
@@ -39,28 +63,271 @@ public abstract class ValidateTransportVersionDefinitionsTask extends DefaultTas
+            }
+        }
+    }
+
+    private TransportVersionDefinition readExistingDefinition(String name) {
+        return readExistingFile(name, this::definitionRelativePath, TransportVersionDefinition::fromString);
+    }
+
+    private TransportVersionLatest readExistingLatest(String branch) {
+        return readExistingFile(branch, this::latestRelativePath, TransportVersionLatest::fromString);
+    }
+
+    private <T> T readExistingFile(String name, Function<String, String> pathFunction, BiFunction<String, String, T> parser) {
+        String relativePath = pathFunction.apply(name);
+        if (existingResources.contains(relativePath) == false) {
+            return null;
+        }
+        String content = gitCommand("show", "main:" + relativePath).strip();
+        return parser.apply(relativePath, content);
+    }
+
+    private void recordAndValidateLatest(TransportVersionLatest latest) {
+        latestByBranch.put(latest.branch(), latest);
+
+        TransportVersionDefinition latestDefinition = definitions.get(latest.name());
+        if (latestDefinition == null) {
+            throwLatestFailure(latest.branch(), "contains transport version name [" + latest.name() + "] which is not defined");
+        }
+        if (latestDefinition.ids().contains(latest.id()) == false) {
+            throwLatestFailure(
+                latest.branch(),
+                "has id " + latest.id() + " which is not in definition [" + definitionRelativePath(latest.name()) + "]"
+            );
+        }
+
+        List<IdAndDefinition> baseIds = idsByBase.get(latest.id().base());
+        IdAndDefinition lastId = baseIds.getLast();
+        if (lastId.id().complete() != latest.id().complete()) {
+            throwLatestFailure(
+                latest.branch(),
+                "has id "
+                    + latest.id()
+                    + " from ["
+                    + latest.name()
+                    + "] with base "
+                    + latest.id().base()
+                    + " but another id "
+                    + lastId.id().complete()
+                    + " from ["
+                    + lastId.definition().name()
+                    + "] is later for that base"
+            );
+        }
+
+        TransportVersionLatest existingLatest = readExistingLatest(latest.branch());
+        if (existingLatest != null) {
+            if (latest.id().patch() != 0 && latest.id().base() != existingLatest.id().base()) {
+                throwLatestFailure(latest.branch(), "modifies base id from " + existingLatest.id().base() + " to " + latest.id().base());
+            }
+        }
+    }
+
+    private void cleanupAndValidateBase(int base, List<IdAndDefinition> ids) {
+        // first sort the ids list so we can check compactness and quickly lookup the highest id later
+        ids.sort(Comparator.comparingInt(a -> a.id().complete()));
```

Comment:
Do we actually want to sort these ids, or should we fail validation if they are not sorted? Perhaps this logic should be in the code that gets the ids initially, avoiding mutation?